pkg/vault/contracts/AddUnbalancedLiquidityViaSwapRouter.sol

```diff
@@ -0,0 +1,228 @@
+
+    function queryAddUnbalancedLiquidityViaSwapExactIn(
+        address pool,
+        address sender,
+        AddLiquidityProportionalParams calldata addLiquidityParams,
+        SwapExactInParams calldata swapParams
+    ) external saveSender(sender) returns (uint256[] memory amountsIn, uint256 swapAmountOut) {
+        (amountsIn, swapAmountOut) = abi.decode(
+            _vault.quote(
+                abi.encodeCall(
+                    AddUnbalancedLiquidityViaSwapRouter.queryAddUnbalancedLiquidityViaSwapHook,
+                    _buildAddLiquidityParams(
+                        pool,
+                        address(this),
+                        _MAX_AMOUNT, // No deadline in query
+                        false, // wethIsEth is false in query
+                        addLiquidityParams,
+                        _buildSwapExactInParams(swapParams)
+                    )
+                )
+            ),
+            (uint256[], uint256)
+        );
+    }
+
+    function queryAddUnbalancedLiquidityViaSwapExactOut(
+        address pool,
+        address sender,
+        AddLiquidityProportionalParams calldata addLiquidityParams,
+        SwapExactOutParams calldata swapParams
+    ) external saveSender(sender) returns (uint256[] memory amountsIn, uint256 swapAmountIn) {
+        (amountsIn, swapAmountIn) = abi.decode(
+            _vault.quote(
+                abi.encodeCall(
+                    AddUnbalancedLiquidityViaSwapRouter.queryAddUnbalancedLiquidityViaSwapHook,
+                    _buildAddLiquidityParams(
+                        pool,
+                        address(this),
+                        _MAX_AMOUNT, // No deadline in query
+                        false, // wethIsEth is false in query
+                        addLiquidityParams,
+                        _buildSwapExactOutParams(swapParams)
+                    )
+                )
+            ),
+            (uint256[], uint256)
+        );
+    }
+
+    /***************************************************************************
+                                   Hooks
+    ***************************************************************************/
+
+    function addUnbalancedLiquidityViaSwapHook(
+        AddLiquidityAndSwapHookParams calldata params
+    ) external nonReentrant onlyVault returns (uint256[] memory amountsIn, uint256 swapAmountCalculated) {
+        (amountsIn, , ) = _addLiquidityHook(params.addLiquidityParams);
+        swapAmountCalculated = _swapSingleTokenHook(params.swapParams);
+    }
+
+    function queryAddUnbalancedLiquidityViaSwapHook(
+        AddLiquidityAndSwapHookParams calldata params
+    ) external nonReentrant onlyVault returns (uint256[] memory amountsIn, uint256 swapAmountCalculated) {
+        (amountsIn, , ) = _queryAddLiquidityHook(params.addLiquidityParams);
+        swapAmountCalculated = _querySwapHook(params.swapParams);
+    }
```

Comment:
I'm a bit on the fence with this interface. It's better than before, but I'm leaning towards one more twist unless you think I'm missing something.

From the outside this is just a way to add liquidity, so the swap amount calculated doesn't really matter.
Perhaps it's fine to return these values in the internal hooks, but maybe I'd modify the external interface. See comment below.